SG14_test/slot_map_test.cpp

```diff
@@ -446,6 +446,25 @@ void BoundsCheckingTest()
     assert(cit == sm.end());
 }
 
+template<class SM>
+void GenerationTest()
+{
+    using T = typename SM::mapped_type;
+    SM sm;
+
+    for (int i = 0; i < 10; ++i) {
+        auto k = sm.insert(Monad<T>::from_value(i));
+#if __cplusplus >= 201703L
+        auto[idx, gen] = k;
+        assert(int(gen) == i);
```

Comment:
I feel like this is testing specifically behavior that we _don't_ want (even though my code currently has this behavior). If we change the free-list from FIFO to LIFO, then this test rightly fails, right?

I wrote a test locally that looks like this:

```
template<class SM>
static void GenerationsDontSkipTest()
{
    using T = typename SM::mapped_type;
    SM sm;
    auto k = sm.emplace(Monad<T>::from_value(0));
    auto original_cap = sm.slot_count();
    for (int i=1; i < original_cap; ++i) {
        sm.emplace(Monad<T>::from_value(i));
    }
    assert(sm.size() == sm.slot_count());

    sm.erase(k);
    auto k2 = sm.emplace(Monad<T>::from_value(1));
#if __cplusplus >= 201701L
    auto [idx, gen] = k;
    auto [idx2, gen2] = k2;
    assert(idx2 == idx);
    assert(gen2 == gen + 1);
#else
    using std::get;
    assert(get<0>(k2) == get<0>(k));
    assert(get<1>(k2) == get<1>(k) + 1);
#endif
}
```
That is, we use up all the available indices, and then erase-and-insert one element. We know its index can't change (because it's the only available index), so its generation must increase by 1.

On second thought, I don't think my original idea (writing a test that causes the generation count to wrap around) is terribly good.

Reply:
In theory, at least how I think one would implement LIFO, my test would still pass. A LIFO would still priorities reusing keys vs. creating new indexes, for memory reasons. So adding and removing 1 element should increment generation by 1, because it is the "last in" key. However, if you choose some more complex memory vs. reuse heuristic, then yes it would fail.

I'll add your test as well, but I do want to keep the original (I like it). More tests the merrier. We can deal with LIFO when that comes around.